src/couch_views/src/couch_views_indexer.erl

```diff
@@ -258,7 +260,24 @@ do_update(Db, Mrst0, State0) ->
                     update_stats := UpdateStats
                 }}
         end
-    end).
+    end),
+    case TxResult of
+        {Mrst, continue, State} ->
+            {Mrst, State};
+        {Mrst, finished, State} ->
+            #{tx_db := OldDb} = State,
+            CommittedVer = erlfdb:get_committed_version(maps:get(tx, OldDb)),
+            fabric2_fdb:transactional(OldDb#{tx := undefined}, fun(TxDb) ->
```

Comment:
That could work. Good idea. We already have a recent GRV so we could use it. However, since this will be a write transaction there is a risk to be stuck in a conflict loop here (job data values could be fairly contended). But we could set it up as an opportunistic GRV set, such that we try once only then ignore it on an error retry (with a pdict trick).

Reply:
Implemented in a [fixup]( https://github.com/apache/couchdb/pull/3391/commits/c29f5b24437e0cddc1632586084b4698ff175edd)

`erlfdb:get_last_error/0` made it quite easy to check if the transactional closure is being retried due to an error.

Reply:
So I agree we want to avoid getting into an infinite retry loop if the committed version is unusable.

Are edit conflicts a real possibility here? The `set_build_status` function is doing a blind write on the key associated with this view group, so it will never conflict on its own. But if there is some contention on the view group then I worry that mutating this key blindly could have issues, e.g. could we set it to `?INDEX_READY` in a retry after another transaction (e.g. the next view build) has configured it differently?

The code in `couch_jobs_fdb:update/3` is a bit too dense for me to follow easily. Would there possibly be other contended updates on the key range associated with this job? Or is this transaction just reading other keys that could be updated elsewhere and bumping into conflicts that way ...

Reply:
Since we're re-using the GRV from the last `do_update` call, another transaction might have read something that we forgot to snapshot in our `do_update` transaction and then did any writes before this second finalize transaction ran. Though we do read the changes and doc bodies with snapshots now, so that leaves the `couch_views_fdb:get_view_state/2` call but maybe others (or news ones added in the future).. That would be a spurious (un-intended conflict) in way and we'd be stuck in a loop.

> could we set it to ?INDEX_READY in a retry after another transaction (e.g. the next view build) has configured it differently?

Good question. Build state should be set by the indexer only. And only one indexer may operate on the job at the time. That's exactly what the `couch_jobs_fdb:update/3` accomplishes. The logic in update/3 as follows:

 * `get` the job data and state
 * If the lock uuid doesn't match the one read, then exit with `halt`. This means someone else might have acquired the job or it was deleted, etc.
 * Otherwise, `put` the job data and then bump the activity (both the watch for the whole job type and the individual jobs)

So nobody else should be clobbering the index_ready state, as long as the current job is holding the job lock. And it may hold the lock across multiple transactions. For that to work though, indexers _must_ call `couch_jobs:update/3` every time they update the index (that happens in `report_progress/2`).  Which means in the current PR there is a bug because during the last do_update call we remove the call `report_progress/2`(I thought it would be an optimization, ha!). Good eye, @kocolosk!

It seems then we need to ensure we call progress update in the current PR on the last do_update calls, but maybe also keep the opportunistic GRV set to prevent possible spurious conflicts.

Reply:
Fixed in d5fae35f695758a2b21bf8936f28e5c310f894c4

Reply:
You're giving me entirely too much credit on that one, I had absolutely no idea about the `couch_jobs:update/3` contract :)

Reply:
You had the right intuition to ask about contended updates for the key range of that job!